Codegen/src/Nodes/TypedDataNode.cs

```diff
@@ -924,19 +926,97 @@ private static System.Text.StringBuilder GetIndent(int level, int indent, bool i
+            List<CopyDirective> usedCopies = new List<CopyDirective>();
             foreach (var child in type.Children)
             {
-                bool bCanGenerate = !child.IsInsideCopy();
+                bool bIsInsideCopy = child.IsInsideCopy();
+
                 //Handle Typedef whose body is inside a COPY
                 if (child.IsFlagSet(Flag.InsideTypedefFromCopy))
                 {
                     if (child.IsFlagSet(Flag.IsTypedefCopyNode))
                     {
-                        lines.AddRange(child.Lines);
+                        if (child.Lines.Sum(l => l.Text.Count(c => c == '.')) > 1)
+                        {
+                            //If there is more than 1 instruction displayed in the lines, create new line containing only the relevant text
+                            var lineText = child.Lines.First().Text.GetIndent() + child.CodeElement.Text;
+                            lines.Add(new CobolTextLine(new TextLineSnapshot(child.CodeElement.Line, lineText, null), ColumnsLayout.CobolReferenceFormat));
+                        }
+                        else
+                        {
+                            lines.AddRange(child.Lines);
+                        }
+
                         continue;
                     }
                 }
 
+                //If the child is coming from a copy, we want to add the clause copy to the lines
+                if (bIsInsideCopy && child.CodeElement != null)
+                {
+                    //There is already a mechanism to write the clause COPY coming from the main program in LinearNodeSourceCodeMapper. 
+                    //This is how we recover the clause copy from the dependencies.
+                    //Here we check if the typedef has been defined in another file than the one declaring the datatype.
+                    if (type.Root != null && !type.Root.Programs.Any(p => rootProcedures.Contains(p.Name)))
+                    {
+                        CopyDirective copy = child.CodeElement.FirstCopyDirective;
+                        //The first data coming from a copy is used to recover the Clause COPY, the other would be only a repetition of this one, so we skip them.
+                        //Even with the same name, two different Clause COPY are differentiated by their token lines
+                        if (usedCopies.Contains(copy) == false)
+                        {
+                            //We recreate the clause copy with its arguments.
+                            //We don't have access to the original tokens here, so we can't recreate the same format as the original sentence.
+                            StringBuilder copyInstruction = new StringBuilder();
+
+                            copyInstruction.Append(copy.COPYToken.TokensLine.SourceText.GetIndent());
+                            copyInstruction.Append("COPY " + copy.TextName);
+                            if (copy.ReplaceOperations.Any())
```

Comment:
Thanks to @fm-117 all this code between line 973 and 1010  can be replaced by a code like this:

                            copyInstruction.Append(copy.COPYToken.TokensLine.SourceText.GetIndent());
                            copy.ConsumedTokens.SelectedTokensOnSeveralLines.All(tokenLine =>
                            {
                                tokenLine.All(t =>
                                {
                                    copyInstruction.Append(t.Text);
                                    return true;
                                });
                                return true;
                            });

Reply:
But It will be necessaty to add as many lines as there are elements in copy.ConsumedTokens.SelectedTokensOnSeveralLines maybe

Reply:
See last commits